CMakeLists.txt

```
@@ -1,170 +1,190 @@
-  KOKKOSKERNELS_ADD_TEST_DIRECTORIES(sparse/unit_test)
-  KOKKOSKERNELS_ADD_TEST_DIRECTORIES(ode/unit_test)
-  KOKKOSKERNELS_ADD_TEST_DIRECTORIES(perf_test)
-  KOKKOSKERNELS_ADD_EXAMPLE_DIRECTORIES(example)
-ELSE()
+  kokkoskernels_add_test_directories(common/unit_test)
+  kokkoskernels_add_test_directories(batched/dense/unit_test)
+  kokkoskernels_add_test_directories(batched/sparse/unit_test)
+  kokkoskernels_add_test_directories(blas/unit_test)
+  kokkoskernels_add_test_directories(lapack/unit_test)
+  kokkoskernels_add_test_directories(graph/unit_test)
+  kokkoskernels_add_test_directories(sparse/unit_test)
+  kokkoskernels_add_test_directories(ode/unit_test)
+  kokkoskernels_add_test_directories(perf_test)
+  kokkoskernels_add_example_directories(example)
+else()
   # Regular build, not install testing
   # Do all the regular option processing
-  IF (NOT KOKKOSKERNELS_HAS_TRILINOS AND NOT KOKKOSKERNELS_HAS_PARENT)
+  if (NOT KOKKOSKERNELS_HAS_TRILINOS AND NOT KOKKOSKERNELS_HAS_PARENT)
     # This is a standalone build
-    FIND_PACKAGE(Kokkos REQUIRED)
-    IF(Kokkos_COMPILE_LANGUAGE)
-      ENABLE_LANGUAGE(${Kokkos_COMPILE_LANGUAGE})
-      IF(Kokkos_COMPILE_LANGUAGE STREQUAL CUDA)
-        SET(CMAKE_CUDA_ARCHITECTURES ${Kokkos_CUDA_ARCHITECTURES})
-      ENDIF()
-      IF(Kokkos_COMPILE_LANGUAGE STREQUAL HIP)
-        SET(CMAKE_HIP_ARCHITECTURES ${Kokkos_HIP_ARCHITECTURES})
-      ENDIF()
-    ENDIF()
-    IF(${Kokkos_VERSION} VERSION_GREATER_EQUAL "4.6.02")
-      MESSAGE(STATUS "Found Kokkos version ${Kokkos_VERSION} at ${Kokkos_DIR}")
-      IF((${Kokkos_VERSION} VERSION_GREATER "4.7.99"))
-        MESSAGE(WARNING "Configuring with Kokkos ${Kokkos_VERSION} which is newer than the expected develop branch - version check may need update")
-      ENDIF()
-    ELSE()
-      MESSAGE(FATAL_ERROR "Kokkos Kernels ${KokkosKernels_VERSION} requires Kokkos 4.6.02 or greater (found ${Kokkos_VERSION})")
-    ENDIF()
-  ENDIF()
-
-  INCLUDE(cmake/kokkos_backends.cmake)
+    find_package(Kokkos REQUIRED)
+    if (Kokkos_COMPILE_LANGUAGE)
+      enable_language(${Kokkos_COMPILE_LANGUAGE})
+      if (Kokkos_COMPILE_LANGUAGE STREQUAL CUDA)
+        set(CMAKE_CUDA_ARCHITECTURES ${Kokkos_CUDA_ARCHITECTURES})
+      endif()
+      if (Kokkos_COMPILE_LANGUAGE STREQUAL HIP)
+        set(CMAKE_HIP_ARCHITECTURES ${Kokkos_HIP_ARCHITECTURES})
+      endif()
+    endif()
+    if (${Kokkos_VERSION} VERSION_GREATER_EQUAL "4.6.02")
+      message(STATUS "Found Kokkos version ${Kokkos_VERSION} at ${Kokkos_DIR}")
+      if (${Kokkos_VERSION} VERSION_GREATER "4.7.99")
+        message(WARNING "Configuring with Kokkos ${Kokkos_VERSION} which is newer than the expected develop branch - version check may need update")
+      endif()
+    else()
+      message(FATAL_ERROR "Kokkos Kernels ${KokkosKernels_VERSION} requires Kokkos 4.6.02 or greater (found ${Kokkos_VERSION})")
+    endif()
+
+    if (KokkosKernels_ENABLE_COMPILER_WARNINGS)
```

I might be wrong but aren't you supposed to write cmake logic using the cache variable instead of the user passed parameter?

```suggestion
    if (KOKKOSKERNELS_ENABLE_COMPILER_WARNINGS)
```

Note that I could be wrong here, so please do not just make the change, it would be good to understand what is the proper way to do this

---

`KokkosKernels_XXX` is the CACHE variable:

```
  SET(CAMEL_NAME KokkosKernels_${SUFFIX})
  STRING(TOUPPER ${CAMEL_NAME} UC_NAME)

  # Make sure this appears in the cache with the appropriate DOCSTRING                                                                                                                                                          
  SET(${CAMEL_NAME} ${DEFAULT} CACHE ${TYPE} ${DOCSTRING})
```

There's a lot of stuff in `kokkoskernels_add_option` that doesn't make sense to me and is not standard CMake. I assume this has something to do with tribits. It looks like we make sure that the cache is always camel case but then set non-cache variables in all caps. I assume this is to allow case-insensitive cmake code. In other words, both

`if (KokkosKernels_XXX)` and `if (KOKKOSKERNELS_XXX)` will work. I think it's better to just use the CACHE variable (which is also what the user would pass).